Source/WebCore/platform/network/haiku/BUrlProtocolHandler.cpp

```diff
@@ -46,152 +47,357 @@ static const int gMaxRecursionLimit = 10;
+
+    if (statusCode == 303)
+        return true;
+
+    if ((statusCode == 301 || statusCode == 302) && request.httpMethod() == "POST")
+        return true;
 
-    BString method = BString(m_resourceHandle->firstRequest().httpMethod());
+    if (crossOrigin && request.httpMethod() == "DELETE")
+        return true;
 
-    m_postData = NULL;
+    return false;
+}
+
+RefPtr<BUrlRequestWrapper> BUrlRequestWrapper::create(BUrlProtocolHandler* handler, NetworkStorageSession* storageSession, ResourceRequest& request)
+{
+    return adoptRef(*new BUrlRequestWrapper(handler, storageSession, request));
+}
+
+BUrlRequestWrapper::BUrlRequestWrapper(BUrlProtocolHandler* handler, NetworkStorageSession* storageSession, ResourceRequest& request)
+    : BUrlProtocolAsynchronousListener(true)
+    , m_handler(handler)
+    , m_receiveMutex()
+{
+    ASSERT(isMainThread());
+    ASSERT(m_handler);
+    ASSERT(storageSession);
 
-    if (m_request == NULL)
+    m_request = request.toNetworkRequest(&storageSession->platformSession());
+
+    if (!m_request)
         return;
 
-    m_baseUrl = URL(m_request->Url());
+    m_request->SetListener(SynchronousListener());
+    m_request->SetOutput(this);
 
     BHttpRequest* httpRequest = dynamic_cast<BHttpRequest*>(m_request);
-    if(httpRequest) {
-        // TODO maybe we have data to send in other cases ?
-        if(method == B_HTTP_POST || method == B_HTTP_PUT) {
-            FormData* form = m_resourceHandle->firstRequest().httpBody();
-            if(form) {
-                m_postData = new BFormDataIO(form, context->storageSession()->sessionID());
-                httpRequest->AdoptInputData(m_postData, m_postData->Size());
+    if (httpRequest) {
+        if (request.httpMethod() == "POST" || request.httpMethod() == "PUT") {
+            if (request.httpBody()) {
+                auto postData = new BFormDataIO(request.httpBody(), storageSession->sessionID());
+                httpRequest->AdoptInputData(postData, postData->Size());
             }
         }
 
-        httpRequest->SetMethod(method.String());
+        httpRequest->SetMethod(request.httpMethod().utf8().data());
+        // Redirections will be handled by this class.
+        httpRequest->SetFollowLocation(false);
+    } else if (request.httpMethod() != "GET") {
+        // Only the HTTP backend support things other than GET.
+        // Remove m_request to signify to ResourceHandle that the request was
+        // invalid.
+        delete m_request;
+        m_request = NULL;
+        return;
     }
 
-    // In synchronous mode, call this listener directly.
-    // In asynchronous mode, go through a BMessage
-    if(this->SynchronousListener()) {
-        m_request->SetListener(this->SynchronousListener());
-    } else {
-        m_request->SetListener(this);
-    }
+    // Keep self alive while BUrlRequest is running as we hold
+    // the main dispatcher.
+    ref();
+
+    // Block the receiving thread until headers are parsed.
```

Comment:
I don't remember why I added the lock... Upon a quick inspection it might not be needed, as the lambda produced by `Write()` actually check for `m_handler` validity before acting on the data.

I might have added it because earlier revisions of the patch have `BUrlRequestWrapper::Write()` accessing `m_handler`.